engine/src/state_chain_observer/client.rs

```diff
@@ -265,66 +284,137 @@ impl<BaseRpcClient: base_rpc_api::BaseRpcApi + Send + Sync + 'static, SignedExtr
-					)))
-					.chain(block_stream.by_ref())
-					.try_take_while(|block_info| {
-						let base_rpc_client = base_rpc_client.clone();
-						let block_info = *block_info;
-						async move {
-							Ok(
-								if let Err(block_version) = base_rpc_client
-									.check_block_compatibility(block_info.hash)
+
+			let block_stream: Pin<
+				Box<
+					dyn TryCachedStream<
+							Error = anyhow::Error,
+							Ok = BlockInfo,
+							Item = Result<BlockInfo, anyhow::Error>,
+						> + Send,
+				>,
+			> = if let Some(start_from) = start_from {
+				let base_rpc_client = base_rpc_client.clone();
+
+				let mut start_from_headers = Box::pin(
+					futures::stream::iter(start_from..latest_block.number)
+						.then(move |block_number| {
+							let base_rpc_client_c = base_rpc_client.clone();
+							async move {
+								let hash = base_rpc_client_c
+									.block_hash(block_number)
 									.await?
-								{
-									info!(
-										"{} WAITING for a compatible release version.",
-										incompatible_block_message(block_version, block_info)
-									);
-									true
-								} else {
-									false
-								},
-							)
-						}
-					})
-					.boxed();
-
-				// Note underlying stream ends in Error, therefore it is guaranteed try_for_each
-				// will not exit successfully until a compatible block is found
-				incompatible_blocks
-					.try_for_each(move |_| futures::future::ready(Ok::<_, anyhow::Error>(())))
-					.await?;
+									.ok_or_else(|| {
+										anyhow::anyhow!(
+											"Block number {} does not exist in the state chain",
+											block_number
+										)
+									})?;
+								Ok(base_rpc_client_c.block_header(hash).await?)
+							}
+						})
+						.map_ok(|header| -> BlockInfo { header.into() }),
+				);
+
+				let first_block: BlockInfo = start_from_headers.next().await.unwrap()?;
+				Box::pin(
+					start_from_headers
+						.chain(futures::stream::once(futures::future::ready(
+							Ok::<_, anyhow::Error>(latest_block),
+						)))
+						.chain(block_stream)
+						.make_try_cached(first_block),
+				)
 			} else {
-				base_rpc_client.check_block_compatibility(latest_block.hash).await?.map_err(
-					|latest_block_version| {
-						anyhow::anyhow!(
-							"{}",
-							incompatible_block_message(latest_block_version, latest_block)
-						)
-					},
-				)?;
-			}
+				Box::pin(block_stream.make_try_cached(latest_block))
+			};
+
+			let first_block = *block_stream.cache();
+
+			let mut block_stream = process_stream_fn(block_stream).await?;
 
-			block_stream
+			let (cfe_compatibility, latest_version_runtime_requires) =
+				base_rpc_client.check_block_compatibility(first_block.hash).await?;
+
+			match cfe_compatibility {
+				CfeCompatibility::NoLongerCompatible => {
+					return Err(CreateStateChainClientError::NoLongerCompatible {
+						cfe_version: *CFE_VERSION,
+						cfe_version_required: latest_version_runtime_requires,
+						first_incompatible_block: first_block.number,
```

Comment:
This is not really the `first_imcompatible_block`, I think maybe renaming to `at_block`, or something is more accurate. Because the version could have been changes many blocks ago.

Reply:
fair